### PR TITLE
Port RTCPacketOptions to the new IPC serialization format

### PR DESCRIPTION
**Pull request by cdumez**
#### 58712f35351470bfd43a0ea5a4faefcd780315cf
<pre>
Port RTCPacketOptions to the new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=264779">https://bugs.webkit.org/show_bug.cgi?id=264779</a>

Reviewed by Alex Christensen.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/RTCPacketOptions.cpp:
(WebKit::RTCPacketOptions::RTCPacketOptions):
(WebKit::RTCPacketOptions::serializableData const):
(WebKit::RTCPacketOptions::encode const): Deleted.
(WebKit::RTCPacketOptions::decode): Deleted.
* Source/WebKit/Shared/RTCPacketOptions.h:
* Source/WebKit/Shared/RTCPacketOptions.serialization.in: Added.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/270713@main">https://commits.webkit.org/270713@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/25543665098331e27eb0d8afb8eb447ce1225655

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26103 "21 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27376 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28199 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23901 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2127 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23946 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26358 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3568 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22472 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28775 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3179 "Passed tests") | | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29478 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23798 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23812 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27366 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3222 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1421 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4626 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3687 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3372 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3543 "Built successfully") | | |

### CODE REVIEW

File: Source/WebKit/Shared/RTCPacketOptions.h [hunk 1]
```diff
@@ -28,25 +28,34 @@
 #if USE(LIBWEBRTC)
 
 #include <WebCore/LibWebRTCMacros.h>
-#include <optional>
 #include <wtf/Forward.h>
+#include <wtf/Vector.h>
```
Comment (achristensen07) [inline Source/WebKit/Shared/RTCPacketOptions.h]: This doesn't look used.

File: Source/WebKit/Shared/RTCPacketOptions.h [hunk 2]
```diff
@@ -28,25 +28,34 @@
...
+    struct SerializableData {
+        int32_t dscp;
+        int32_t packetId;
+        int rtpSendtimeExtensionId;
+        int64_t srtpAuthTagLength;
+        std::span<const uint8_t> srtpAuthKey;
+        int64_t srtpPacketIndex;
+    };
+
+    explicit RTCPacketOptions(const rtc::PacketOptions& options)
```
Comment (achristensen07) [inline Source/WebKit/Shared/RTCPacketOptions.h:77]: Can we use a move constructor?

File: Source/WebKit/Shared/RTCPacketOptions.h [hunk 3]
```diff
@@ -28,25 +28,34 @@
...
-class Encoder;
-}
-
 namespace WebKit {
 
 struct RTCPacketOptions {
-    void encode(IPC::Encoder&) const;
-    static std::optional<RTCPacketOptions> decode(IPC::Decoder&);
+    struct SerializableData {
+        int32_t dscp;
```
Comment (achristensen07) [inline Source/WebKit/Shared/RTCPacketOptions.h]: It looks like there are only 23 valid values here.  We should probably use an enum class.

File: Source/WebKit/Shared/RTCPacketOptions.cpp [hunk 4]
```diff
@@ -34,66 +34,35 @@
...
-        if (!decoder.decode(srtpAuthKey))
-            return std::nullopt;
-
-        params.srtp_auth_key = std::vector<char>(static_cast<size_t>(srtpAuthKey.size()));
-        memcpy(params.srtp_auth_key.data(), reinterpret_cast<const char*>(srtpAuthKey.data()), srtpAuthKey.size() * sizeof(char));
+    params.rtp_sendtime_extension_id = data.rtpSendtimeExtensionId;
+    params.srtp_auth_tag_len = data.srtpAuthTagLength;
+    if (data.srtpAuthTagLength > 0) {
+        auto* begin = reinterpret_cast<const char*>(data.srtpAuthKey.data());
+        params.srtp_auth_key = std::vector<char>(begin, begin + static_cast<size_t>(data.srtpAuthKey.size()));
```
Comment (achristensen07) [inline Source/WebKit/Shared/RTCPacketOptions.cpp]: If we make srtpAuthKey a span<const char> we may be able to just use begin() and end() here.

File: Source/WebKit/Shared/RTCPacketOptions.cpp [hunk 5]
```diff
@@ -34,66 +34,35 @@
...
 }
 
+auto RTCPacketOptions::serializableData() const -> SerializableData
+{
+    return {
+        static_cast<int32_t>(options.dscp),
+        safeCast<int32_t>(options.packet_id),
+        options.packet_time_params.rtp_sendtime_extension_id,
+        static_cast<int64_t>(options.packet_time_params.srtp_auth_tag_len),
+        options.packet_time_params.srtp_auth_tag_len > 0  ? std::span<const uint8_t> { } : std::span<const uint8_t> { reinterpret_cast<const uint8_t*>(options.packet_time_params.srtp_auth_key.data()), options.packet_time_params.srtp_auth_key.size() },
```
Comment (achristensen07) [inline Source/WebKit/Shared/RTCPacketOptions.cpp]: It is suspicious that srtp_auth_tag_len is serialized separately, then the length of this is not directly related to that value.  Do we need a bounds check when decoding srtp_auth_tag_len?

**Review by achristensen07 — CHANGES_REQUESTED** [review]

File: Source/WebKit/Shared/RTCPacketOptions.h (hunk 1 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.h]: Indeed, will drop.

File: Source/WebKit/Shared/RTCPacketOptions.h (hunk 2 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.h:77]: Concurrently the call site is being given a `const rtc::PacketOptions&` so not easily. Maybe if I modify the call-site's call site. I'd have to look into it. At this this is not worse than before.

File: Source/WebKit/Shared/RTCPacketOptions.h (hunk 3 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.h]: Ok. Aren't we worried about the maintenance cost given that we'll need to match an enum from a third-party lib?

File: Source/WebKit/Shared/RTCPacketOptions.cpp (hunk 5 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.cpp]: I'm trying to match the behavior of the previous encoder / decoder.

File: Source/WebKit/Shared/RTCPacketOptions.cpp (hunk 4 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.cpp]: srtpAuthKey is a std::span. However, it contains `const uint8_t` instead of `char`.

File: Source/WebKit/Shared/RTCPacketOptions.h (hunk 2 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.h:77]: The call site is:
```
int LibWebRTCSocket::SendTo(const void *value, size_t size, const rtc::SocketAddress& address, const rtc::PacketOptions& options);
```

It takes options as `const rtc::PacketOptions&`. The interface is defined by AsyncPacketSocket in the libWebRTC library so it is not something I can change. I can't pass an r-value here.

File: Source/WebKit/Shared/RTCPacketOptions.h (hunk 1 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.h]: Fixed in latest iteration.

File: Source/WebKit/Shared/RTCPacketOptions.h (hunk 3 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.h]: Fixed in the latest iteration.

File: Source/WebKit/Shared/RTCPacketOptions.cpp (hunk 4 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.cpp]: Fixed in the latest iteration. I am now using a std::span<const char> and using begin() / end() here.

File: Source/WebKit/Shared/RTCPacketOptions.cpp (hunk 5 above)
Comment (cdumez) [reply Source/WebKit/Shared/RTCPacketOptions.cpp]: I don't think there is a security concern here. While we use srtp_auth_tag_len to decide whether or not to decode the span, the span has its own side and its decoding doesn't rely on srtp_auth_tag_len. I think matching the old coders like I did is OK here?

**Review by achristensen07 — APPROVED** [review]
